File: pi_collect.f90
PROGRAM PI_collect

USE mpi_f08
IMPLICIT NONE

DOUBLE PRECISION PI25, sum, pi
INTEGER rank, size, n, i

PARAMETER (PI25 = 3.141592653589793d0)

    CALL MPI_Init()
    CALL MPI_Comm_rank( MPI_COMM_WORLD, rank)
    CALL MPI_Comm_size( MPI_COMM_WORLD, size)

    DO
        IF( rank == 0 ) THEN
            WRITE(*,*) 'Enter the number of intervals (0 quits) : '
            READ(*,*) n
        ENDIF

        CALL MPI_Bcast( ... )
        IF ( n <= 0 ) EXIT

        sum = 0.0d0
        DO i = n - 1 - rank, 0, -size
            sum = sum + (4.0d0 - 8 * MOD(i, 2)) / (2 * DBLE(i) + 1)
        ENDDO

        CALL MPI_Reduce( ... )

        IF (rank == 0) THEN
            WRITE(*, "(' Pi is approximately: ', F19.16, ' Error is: ', F19.16)") pi, ABS(pi - PI25)
        ENDIF
    ENDDO

    CALL MPI_Finalize()

END PROGRAM pi_collect

File: pi_collect.c
#include <mpi.h>
#include <math.h>
#include <stdio.h>

int main(int argc, char *argv[])
{
    int rank, size, n, i;
    double sum, pi;

    MPI_Init( &argc, &argv );
    MPI_Comm_size( MPI_COMM_WORLD, &size );
    MPI_Comm_rank( MPI_COMM_WORLD, &rank );

    do {
        if (rank == 0) {
            printf("Enter the number of intervals (0 quits) : ");
            fflush(stdout);
            scanf("%d", &n);
            if (n > 0) printf("-> %d iterations\n", n); else printf("done\n");
        }

        MPI_Bcast( ... );
        if (n == 0) break;

        sum = 0.0;
        for (i = n - 1 - rank; i >= 0; i -= size) {
            sum += (4.0 - 8.0 * (i % 2)) / (2.0 * i + 1);
        }

        MPI_Reduce( ... );

        if (rank == 0) {
            printf("PI is approximately %.16f, Error is %.16f\n",
                pi, fabs(pi - M_PI));
        }
    } while (1);

    MPI_Finalize();
    return 0;
}

